### PR TITLE
Define timeseries schema in TOML

### PR DESCRIPTION
**Pull request by bnaecker**
This is the first commit in a thread of work to support updates to timeseries schema. In our first step, we're moving the definition of timeseries from the Rust structs we use today, to TOML text files. Each file describes one target and all the metrics associated with it, from which we generate exactly the same Rust code that it replaces.

This opens the door to a lot of downstream work, including well-defined updates to schema; detection of conflicting schema; improved metadata for timeseries; and generation of documentation. As an example, this moves exactly one (1) timeseries into the new format, the physical datalink statistics tracked through illumos kstats.

- Move the `oximeter` crate into a private implementation crate, and re-export it at its original path
- Add intermediate representation of timeseries schema, and code for deserializing TOML timeseries definitions, checking them for conflicts, and handling updates. Note that **no actual updates** are currently supported. The ingesting code includes a check that version numbers are exactly 1, except in tests. We include tests that check updates in a number of ways, but until the remainder of the work is done, we limit timeseries in the wild to their current version of 1.
- Add a macro for consuming timeseries definitions in TOML, and generating the equivalent Rust code. Developers should use this new `oximeter::use_timeseries!()` proc macro to create new timeseries.
- Add an integration test in Nexus that pulls in timeseries schema ingested with with `oximeter::use_timeseries!()`, and checks them all for conflicts. As developers add new timeseries in TOML format, they must be added here as well to ensure we detect problems at CI time.
- Updates `kstat-rs` dep to simplify platform-specific code. This is part of this commit because we need the definitions of the physical-data-link timeseries for our integration test, but those were previously only compiled on illumos platforms. They're now included everywhere, while the tests remain illumos-only.

### CODE REVIEW

**Comment by bnaecker** [comment]
Note to self: make issues for sequence of work, and refer to them in this PR instead of `TODO(ben)`.

File: oximeter/impl/src/schema/ir.rs
Comment (jgallagher) [inline oximeter/impl/src/schema/ir.rs:397]: Tiny nit - maybe I'm Stockholm-syndromed into long errors, but I'd probably not have written this function at all, and inlined all the error returns. If you really like this, it could maybe take `s: Into<String>` to save even more call site boilerplate (in the cases where we pass a static string and not a `format!()`'d one).

File: oximeter/impl/src/schema/ir.rs
Comment (jgallagher) [inline oximeter/impl/src/schema/ir.rs:406]: Tiny nit - can we preserve the full error type here, instead of squashing it into `e.to_string()`? If it's important to keep it a string, maybe consider using `InlineErrorChain::new(&e).to_string()` so we get any upstream errors included?

File: oximeter/impl/src/schema/ir.rs
Comment (jgallagher) [inline oximeter/impl/src/schema/ir.rs:460]: Nit - could we replace `_` with an explicit list of variants, so new variants added in the future have to be considered?

File: oximeter/impl/src/schema/mod.rs
Comment (jgallagher) [inline oximeter/impl/src/schema/mod.rs:166]: These fields being `pub` means a client can create `TimeseriesDescription`s that aren't valid (e.g., a target containing `:`). Is that okay? (Same question for the derived `Deserialize` impl, which will allow any strings)

File: oximeter/impl/src/schema/mod.rs
Comment (jgallagher) [inline oximeter/impl/src/schema/mod.rs:182]: I'm not sure this method is worth it - `Units::count()` is two characters longer than just `Units::Count`

File: oximeter/impl/src/schema/mod.rs
Comment (jgallagher) [inline oximeter/impl/src/schema/mod.rs:163]: Does `Default` make sense here? That would just be an empty description, right?

File: oximeter/impl/src/schema/mod.rs
Comment (jgallagher) [inline oximeter/impl/src/schema/mod.rs:197]: How is this related to `timeseries_name`? If it's expected to always match exactly, could we drop it entirely and just use `timeseries_name` as needed?

File: oximeter/impl/src/schema/mod.rs
Comment (jgallagher) [inline oximeter/impl/src/schema/mod.rs:205]: Why do these fields need `#[serde(default = ..)]` tags? The defaults don't seem super intuitive to me, but maybe that's just my unfamiliarity.

File: oximeter/impl/src/schema/mod.rs
Comment (jgallagher) [inline oximeter/impl/src/schema/mod.rs:362]: Tiny nit - do you think a more descriptive name would be useful here? E.g. based on the comment, maybe `ViewableToAll`?

File: oximeter/impl/src/schema/mod.rs
Comment (jgallagher) [inline oximeter/impl/src/schema/mod.rs:367]: Similar nit as above - not sure this method is worth it

File: oximeter/instruments/schema/physical-data-link.toml
Comment (jgallagher) [inline oximeter/instruments/schema/physical-data-link.toml:7]: Is this leftover from testing, or left intentionally? Maybe add a comment if the latter?

File: oximeter/instruments/src/kstat/mod.rs
Comment (jgallagher) [inline oximeter/instruments/src/kstat/mod.rs:282]: No objection here but just curious about the difference - is there a preference for this over using `clock_gettime()`? (I guess the constant is different for illumos vs linux, which is a little annoying, but it'd be _less_ of a difference?) Or, is there a reason we can't use `Instant` and forego the platform-specific bits altogether?

File: oximeter/oximeter/src/bin/oximeter-schema.rs
Comment (jgallagher) [inline oximeter/oximeter/src/bin/oximeter-schema.rs:89]: Thank you! This is much nicer than using something like `cargo expand` to see the emitted code.

File: oximeter/oximeter/src/lib.rs
Comment (jgallagher) [inline oximeter/oximeter/src/lib.rs:48]: Is it okay for this to come after the `[[metrics]]` section? Or, how do we define fields on the target vs fields on the metric(s)?

**Review by jgallagher — COMMENTED** [review]
Sorry for the many nits and questions / thanks for your patience! I haven't done a lot with oximeter so I hope these aren't way off base.

File: schema/rss-sled-plan.json
Comment (bnaecker) [reply schema/rss-sled-plan.json:198]: This one is out of our handles actually. `schemars` introduced a bug which removed NULL values from optional fields with a default. I updated the `schemars` dep, which pulled in a fix that put it _back_. [Here](https://github.com/GREsau/schemars/issues/295) is the original `schemars` bug report, which links to the fix.

File: oximeter/impl/src/schema/codegen.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/codegen.rs:35]: Yeah, that's a good idea. I hadn't actually used this call much before this PR, but now that it's getting more mileage, adding a struct with names is a good call.

File: oximeter/impl/src/lib.rs
Comment (bnaecker) [reply oximeter/impl/src/lib.rs:44]: Maybe. the `Target::name` and `Metric::name` methods all return the snake_case name of the struct right now (assuming they're derived, which they all are, I checked). So since `:` isn't valid in a Rust identifier, we can't have the names contain that character.

That said, it's definitely _possible_, in that folks could implement the traits themselves and create invalid names. I don't have a strong preference, so would be happy to add that check if you think it's worthwhile.

File: oximeter/impl/src/schema/ir.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/ir.rs:397]: I'm definitely not wedded to this. It was mostly evolution rather than design anyway! I can change it up.

File: oximeter/impl/src/schema/ir.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/ir.rs:406]: IIRC, I tried keeping the error directly but hit some trait bound. I'm not sure what that would have been though, now that I see the error implements the common traits like `Clone` and `Display`. I'll try it again, and possibly use the inline-error chain trick if that doesn't work.

File: oximeter/impl/src/schema/ir.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/ir.rs:460]: Yeah, good call, will update.

File: oximeter/impl/src/schema/mod.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs:166]: The _description_ is just freeform text. The _name_ cannot have `:` in it, but that's checked in other ways. I don't have any reason to restrict the format of the description at this time, except possibly limiting the total length?

File: oximeter/impl/src/schema/mod.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs:182]: I need it for the `#[serde(default = "...")]` attribute, which requires a callable function rather than a variant name, unfortunately. I agree, it's silly.

File: oximeter/impl/src/schema/mod.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs:163]: Correct, it's an empty description, and I think that does make sense. The intent was to support the sequence of changes where we start _adding_ these extra pieces of metadata to existing schema that don't currently have them. This type needs `Default` implemented so we can ask serde to create us a default on deserialization.

This is one of the areas I do appreciate extra eyes on though. I am added default values for deserialization to all the new fields of a schema:

- version (default = 1)
- description (default is empty, which you noted here)
- authz_scope (defaults to fleet)
- units (defaults to count)

I _think_ this is the right step. The plan is to add defaults for all these so that we can send them across serialization boundaries; add them into the CRDB and ClickHouse schema; and then start rewriting schema to use the TOML format where these will be provided with "real" values. But I need the defaults first, to support serialization.

File: oximeter/impl/src/schema/mod.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs:197]: Ah, I see the confusion of some earlier comments. The _description_ is a freeform textual description of what the timeseries is supposed to measure, with a component for the target and metric. That provides extra context and details, beyond the existing _name_. It's a docstring, basically.

File: oximeter/impl/src/schema/mod.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs:205]: I had previously convinced myself that we need defaults for these, to support implementing this whole thing in pieces. I don't think that's right anymore.

Mostly as a note to myself, I've written out why this is the case. The default is useful when the client expects different data from what the server produces. For example, the client expects a `units` field, but the serer doesn't provide one. We're not in that situation. Here, the server produces new fields that the client isn't aware of. Older clients can ignore these new fields. And newer clients which expect this field can only exist once we've deployed the code that produces it. Defaults aren't needed in either case.

Part of my confusion is around the databases. Currently `oximeter` derives a schema for each sample, and inserts it in ClickHouse. But we don't have columns for things like the `units` yet. However, we still don't need a `#[serde(default)]`, because we use the internal `DbTimeseriesSchema` type to model the database field. In the conversion from that to the real schema type, we can insert the default, but that's not a `#[serde(default)]`.

Anyway, sorry for the confusion on this, and thanks for bringing this up again. I'll remove all the default attributes and implementations that I've added here.

File: oximeter/impl/src/schema/mod.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs:362]: Sure, that sounds good.

File: oximeter/impl/src/schema/mod.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs:367]: Same comment -- I used this for `#[serde(default = "...")]`. Since that's going away, I'll remove this field too.

File: oximeter/instruments/schema/physical-data-link.toml
Comment (bnaecker) [reply oximeter/instruments/schema/physical-data-link.toml:7]: That was intentional. I have a check in non-test code that the version number is never greater than 1, so that even though we have code to handle and tests to verify multiple versions, you can't actually _create_ multiple versions right now. I was going to remove that as the very last step.

This is a breadcrumb to show what an update would look like. I'll add a comment to that effect.

File: oximeter/instruments/src/kstat/mod.rs
Comment (bnaecker) [reply oximeter/instruments/src/kstat/mod.rs:282]: Yeah, I considered making the `cfg`-detection be around the clock type. I don't have a strong preference, so that might be more reasonable. `man clock_gettime` is clear that the high-res clock serves as the source for `gethrtime()` anyway, so it would be clearer and simpler to do it that way. Thanks.

File: oximeter/oximeter/src/lib.rs
Comment (bnaecker) [reply oximeter/oximeter/src/lib.rs:48]: In this TOML format, the fields are independent of the metrics and the target, so at the same "level" of the TOML hierarchy. The target and metric definitions refer to the fields by name. So I think they can go before or after, since we deserialize and then look up each field by name in the deserialized type.

File: oximeter/impl/src/schema/mod.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs:163]: My previous comment isn't quite right. We don't actually need defaults for deserialization, so I'm going to remove this trait derivation.

File: openapi/bootstrap-agent.json
Comment (bnaecker) [reply openapi/bootstrap-agent.json:331]: Same thing, this was fixing a previous bug introduced by `schemars`.

File: oximeter/impl/src/schema/codegen.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/codegen.rs]: I opted instead for methods which return the relevant element of the tuple instead. See 5610bb467

File: oximeter/impl/src/schema/mod.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs]: Removed in 5610bb467

File: oximeter/impl/src/schema/mod.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs:191]: I removed the `#[serde(default)]` attributes and the constructors used to create the defaults in 5610bb467

File: oximeter/impl/src/schema/mod.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs:363]: Removed in 5610bb467

File: oximeter/impl/src/schema/mod.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs:358]: Renamed in dca0355f3

File: oximeter/instruments/schema/physical-data-link.toml
Comment (bnaecker) [reply oximeter/instruments/schema/physical-data-link.toml:10]: Added a comment in dca0355f3

File: oximeter/instruments/src/kstat/mod.rs
Comment (bnaecker) [reply oximeter/instruments/src/kstat/mod.rs]: Reworked this in dca0355f3 so only the clock-source is platform-dependent

File: oximeter/impl/src/schema/ir.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/ir.rs]: Done in dca0355f3

File: oximeter/impl/src/schema/ir.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/ir.rs:406]: Ah, this type shows up in our API, so it needs serialization and `JsonSchema` traits. I'll use the error-chain approach you recommend, thanks.

File: oximeter/impl/src/schema/ir.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/ir.rs]: Removed in a410171f4

File: oximeter/impl/src/schema/ir.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/ir.rs]: Done in a410171f4

File: oximeter/impl/src/schema/mod.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs]: D'oh, actually renamed in a410171f4

File: openapi/nexus.json
Comment (bnaecker) [reply openapi/nexus.json:9347]: I think they are, yes. We're adding a bunch of new concepts to the public API, so they need to show up here.

File: oximeter/impl/src/schema/mod.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs:191]: Hmm, so we _do_ actually need the defaults for [this test](https://buildomat.eng.oxide.computer/wg/0/details/01J0PT87YHXSPPPTX7ZXQ3NZCT/m6KJAqWpc4klzlu2Akw78hwN89jbzUThRPeG33pW6mN4Htb5/01J0PT8KF46B6ERA5ZZRBECTQS#S5148). That test will be obviated by a lot of this work here, so I might remove it now. But I will dig into it a bit first.

Comment (bnaecker) [reply oximeter/impl/src/schema/mod.rs:191]: I fixed this by just overwriting the test file to include the new fields, in 13048c1333cd8b4f828a45522fd0f886879be6ec

**Comment by bnaecker** [comment]
Ok, I believe I've addressed most of @jgallagher's comments. The big outstanding chunk I'd like to do is bite the bullet and centralize everything in one crate, probably `oximeter` itself. That will change the `use_timeseries!()` macro, since it'll just refer to the target and possibly metric, rather than the full file path. It will also obviate the integration test I have in Nexus, which I'll replace with one or more similar tests in `oximeter` itself.

File: oximeter/impl/src/lib.rs
Comment (jgallagher) [reply oximeter/impl/src/lib.rs:44]: I guess there are three options:

1. Leave as-is
2. Change this to `try_from()` and unwrap the result
3. Change this to `try_from()` and return an error

I think 1 and 2 are almost the same: derived implementations will be fine, an incorrect manual implementation would panic here in 2, and an incorrect manual implementation would (probably?) panic somewhere else later? 3 would be the safest but might also be annoying since this function is currently infallible.

If 3 is super easy I'd do that, but if it's annoying I might go with option 2, and make it something like `.try_from().expect("illegal name: implementations should be derived")` or whatever to make it clear that the only reason it's failing is because someone went out of their way to do it wrong.

File: oximeter/impl/src/schema/mod.rs
Comment (jgallagher) [reply oximeter/impl/src/schema/mod.rs:165]: Oh, yeah, I completely misunderstood this (which you realized in the subsequent comment, haha). Sorry! Ignore this suggestion.

File: oximeter/oximeter/src/lib.rs
Comment (jgallagher) [reply oximeter/oximeter/src/lib.rs:48]: Ok. I've had weird experiences with table orderings and TOML before, but I can never remember the details.

Comment (bnaecker) [reply oximeter/oximeter/src/lib.rs:48]: That's definitely true. Any array of tables always refers to the most recent entry. So if you have `[[foo]]` and then do `[[foo.bar]]`, that's a sub-table in the most recent entry in the `foo` array. There are probably other situations too, but thankfully they don't apply here!

File: oximeter/impl/src/lib.rs
Comment (bnaecker) [reply oximeter/impl/src/lib.rs]: I made this fallible in d358d61c8

**Comment by bnaecker** [comment]
Ok @jgallagher I've also centralized the timeseries definitions into the `oximeter/oximeter/schema` subdirectory. The `use_timeseries!()` proc-macro now takes the name of a file only, inside that subdirectory, and generates the Rust code from its contained timeseries. We can now also easily run a test in `oximeter` that checks the consistency across all timeseries defined in that directory. That should make it easy to detect conflicts, without developers needing to remember to add files anywhere.

File: oximeter/timeseries-macro/src/lib.rs
Comment (jgallagher) [inline oximeter/timeseries-macro/src/lib.rs:19]: Not sure how much we care about this, but `.join()` is pretty flexible; e.g., if someone calls this with a filename of `/foo/bar/baz`, `SCHEMA_DIRECTORY` will be discarded entirely. If we want to _require_ that this be inside `SCHEMA_DIRECTORY`, we maybe want to call `filename.file_name()` to strip any parent directories or `../` escapes?

**Review by jgallagher — APPROVED** [review]
LGTM! Just one note / question about how strict we want to be with schema location.

File: oximeter/timeseries-macro/src/lib.rs
Comment (bnaecker) [reply oximeter/timeseries-macro/src/lib.rs:19]: Yeah, nice catch, thanks. I might choose to be even more conservative here at this point, just disallowing any slashes in the name at all, and / or checking that the `file_name()` output is the same as the input literal string itself. Thanks!

Comment (bnaecker) [reply oximeter/timeseries-macro/src/lib.rs]: Done in 6a7b8ae7f

**Comment by bnaecker** [comment]
Thanks for the review @jgallagher! All the comments were really helpful. I'm going to give Adam a chance to take a look, and then I'll fix up the conflicts on the way in.

File: oximeter/impl/src/schema/codegen.rs
Comment (ahl) [inline oximeter/impl/src/schema/codegen.rs:121]: Is this right? I *think* that `#( #field_defs ),*` will omit the trailing comma. So if we have one or more items in `field_defs` and then `maybe_datum` is `Some(_)` we'd have:

```rust
pub struct Foo {
    pub x: Y
    pub datum: D
}
```

I'd suggest:


```suggestion
            #( #field_defs, )*
            #maybe_datum ,
```

**Review by ahl — COMMENTED** [review]
maybe I found a bug... or I'm confused

File: oximeter/impl/src/schema/codegen.rs
Comment (bnaecker) [reply oximeter/impl/src/schema/codegen.rs:121]: Nice catch, thanks. This does look like a bug that works by accident. The one set of metrics I'm moving now all have no fields themselves, only those from the target. So we never hit the case you describe. I'll add tests for that stuff now.

Comment (bnaecker) [reply oximeter/impl/src/schema/codegen.rs]: I added tests that catch this and fixed it up in 7cefed7a7

**Comment by bnaecker** [comment]
Aight, I'm going to start rebasing on main to resolve the conflicts before smashing that merge button.

**Pull request merged**
